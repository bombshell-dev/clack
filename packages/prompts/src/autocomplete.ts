import { AutocompletePrompt } from '@clack/core';
import color from 'picocolors';
import {
	type CommonOptions,
	S_BAR,
	S_BAR_END,
	S_CHECKBOX_INACTIVE,
	S_CHECKBOX_SELECTED,
	S_RADIO_ACTIVE,
	S_RADIO_INACTIVE,
	symbol,
} from './common.js';
import { limitOptions } from './limit-options.js';
import type { Option } from './select.js';

function getLabel<T>(option: Option<T>) {
	return option.label ?? String(option.value ?? '');
}

function getFilteredOption<T>(searchText: string, option: Option<T>): boolean {
	if (!searchText) {
		return true;
	}
	const label = (option.label ?? String(option.value ?? '')).toLowerCase();
	const hint = (option.hint ?? '').toLowerCase();
	const value = String(option.value).toLowerCase();
	const term = searchText.toLowerCase();

	return label.includes(term) || hint.includes(term) || value.includes(term);
}

function getSelectedOptions<T>(values: T[], options: Option<T>[]): Option<T>[] {
	const results: Option<T>[] = [];

	for (const option of options) {
		if (values.includes(option.value)) {
			results.push(option);
		}
	}

	return results;
}

interface AutocompleteSharedOptions<Value> extends CommonOptions {
	/**
	 * The message to display to the user.
	 */
	message: string;
	/**
	 * Available options for the autocomplete prompt.
	 */
	options: Option<Value>[] | ((this: AutocompletePrompt<Option<Value>>) => Option<Value>[]);
	/**
	 * Maximum number of items to display at once.
	 */
	maxItems?: number;
	/**
	 * Placeholder text to display when no input is provided.
	 */
	placeholder?: string;
	/**
	 * Validates the value
	 */
	validate?: (value: Value | Value[] | undefined) => string | Error | undefined;
}

export interface AutocompleteOptions<Value> extends AutocompleteSharedOptions<Value> {
	/**
	 * The initial selected value.
	 */
	initialValue?: Value;
	/**
	 * The initial user input
	 */
	initialUserInput?: string;
}

export const autocomplete = <Value>(opts: AutocompleteOptions<Value>) => {
	const prompt = new AutocompletePrompt({
		options: opts.options,
		initialValue: opts.initialValue ? [opts.initialValue] : undefined,
		initialUserInput: opts.initialUserInput,
		filter: (search: string, opt: Option<Value>) => {
			return getFilteredOption(search, opt);
		},
		signal: opts.signal,
		input: opts.input,
		output: opts.output,
		validate: opts.validate,
		render() {
			// Title and message display
			const title = `${color.gray(S_BAR)}\n${symbol(this.state)}  ${opts.message}\n`;
			const userInput = this.userInput;
			const valueAsString = String(this.value ?? '');
			const options = this.options;
			const placeholder = opts.placeholder;
			const showPlaceholder = valueAsString === '' && placeholder !== undefined;

			// Handle different states
			switch (this.state) {
				case 'submit': {
					// Show selected value
					const selected = getSelectedOptions(this.selectedValues, options);
					const label =
						selected.length > 0 ? `  ${color.dim(selected.map(getLabel).join(', '))}` : '';
					return `${title}${color.gray(S_BAR)}${label}`;
				}

				case 'cancel': {
					const userInputText = userInput ? `  ${color.strikethrough(color.dim(userInput))}` : '';
					return `${title}${color.gray(S_BAR)}${userInputText}`;
				}

				default: {
					// Display cursor position - show plain text in navigation mode
					let searchText = '';
					if (this.isNavigating || showPlaceholder) {
						const searchTextValue = showPlaceholder ? placeholder : userInput;
						searchText = searchTextValue !== '' ? ` ${color.dim(searchTextValue)}` : '';
					} else {
						searchText = ` ${this.userInputWithCursor}`;
					}

					// Show match count if filtered
					const matches =
						this.filteredOptions.length !== options.length
							? color.dim(
									` (${this.filteredOptions.length} match${this.filteredOptions.length === 1 ? '' : 'es'})`
								)
							: '';

					// Render options with selection
					const displayOptions =
						this.filteredOptions.length === 0
							? []
							: limitOptions({
									cursor: this.cursor,
									options: this.filteredOptions,
									style: (option, active) => {
										const label = getLabel(option);
										const hint =
											option.hint && option.value === this.focusedValue
												? color.dim(` (${option.hint})`)
												: '';

										return active
											? `${color.green(S_RADIO_ACTIVE)} ${label}${hint}`
											: `${color.dim(S_RADIO_INACTIVE)} ${color.dim(label)}${hint}`;
									},
									maxItems: opts.maxItems,
									output: opts.output,
								});

					// Show instructions
					const instructions = [
						`${color.dim('↑/↓')} to select`,
						`${color.dim('Enter:')} confirm`,
						`${color.dim('Type:')} to search`,
					];

					// No matches message
					const noResults =
						this.filteredOptions.length === 0 && userInput
							? [`${color.cyan(S_BAR)}  ${color.yellow('No matches found')}`]
							: [];

					const validationError =
						this.state === 'error' ? [`${color.yellow(S_BAR)}  ${color.yellow(this.error)}`] : [];

					// Return the formatted prompt
					return [
						`${title}${color.cyan(S_BAR)}`,
						`${color.cyan(S_BAR)}  ${color.dim('Search:')}${searchText}${matches}`,
						...noResults,
						...validationError,
						...displayOptions.map((option) => `${color.cyan(S_BAR)}  ${option}`),
						`${color.cyan(S_BAR)}  ${color.dim(instructions.join(' • '))}`,
						`${color.cyan(S_BAR_END)}`,
					].join('\n');
				}
			}
		},
	});

	// Return the result or cancel symbol
	return prompt.prompt() as Promise<Value | symbol>;
};

// Type definition for the autocompleteMultiselect component
export interface AutocompleteMultiSelectOptions<Value> extends AutocompleteSharedOptions<Value> {
	/**
	 * The initial selected values
	 */
	initialValues?: Value[];
	/**
	 * If true, at least one option must be selected
	 */
	required?: boolean;
}

/**
 * Integrated autocomplete multiselect - combines type-ahead filtering with multiselect in one UI
 */
export const autocompleteMultiselect = <Value>(opts: AutocompleteMultiSelectOptions<Value>) => {
	const formatOption = (
		option: Option<Value>,
		active: boolean,
		selectedValues: Value[],
		focusedValue: Value | undefined
	) => {
		const isSelected = selectedValues.includes(option.value);
		const label = option.label ?? String(option.value ?? '');
		const hint =
			option.hint && focusedValue !== undefined && option.value === focusedValue
				? color.dim(` (${option.hint})`)
				: '';
		const checkbox = isSelected ? color.green(S_CHECKBOX_SELECTED) : color.dim(S_CHECKBOX_INACTIVE);

		if (active) {
			return `${checkbox} ${label}${hint}`;
		}
		return `${checkbox} ${color.dim(label)}`;
	};

	// Create text prompt which we'll use as foundation
	const prompt = new AutocompletePrompt<Option<Value>>({
		options: opts.options,
		multiple: true,
		filter: (search, opt) => {
			return getFilteredOption(search, opt);
		},
		validate: () => {
			if (opts.required && prompt.selectedValues.length === 0) {
				return 'Please select at least one item';
			}
			return undefined;
		},
		initialValue: opts.initialValues,
		signal: opts.signal,
		input: opts.input,
		output: opts.output,
		render() {
			// Title and symbol
			const title = `${color.gray(S_BAR)}\n${symbol(this.state)}  ${opts.message}\n`;

			// Selection counter
			const userInput = this.userInput;
			const placeholder = opts.placeholder;
			const showPlaceholder = userInput === '' && placeholder !== undefined;

			// Search input display
			const searchText =
				this.isNavigating || showPlaceholder
					? color.dim(showPlaceholder ? placeholder : userInput) // Just show plain text when in navigation mode
					: this.userInputWithCursor;

			const options = this.options;

			const matches =
				this.filteredOptions.length !== options.length
					? color.dim(
							` (${this.filteredOptions.length} match${this.filteredOptions.length === 1 ? '' : 'es'})`
						)
					: '';

			// Render prompt state
			switch (this.state) {
				case 'submit': {
					return `${title}${color.gray(S_BAR)}  ${color.dim(`${this.selectedValues.length} items selected`)}`;
				}
				case 'cancel': {
					return `${title}${color.gray(S_BAR)}  ${color.strikethrough(color.dim(userInput))}`;
				}
				default: {
					// Instructions
					const instructions = [
						`${color.dim('↑/↓')} to navigate`,
						`${color.dim('Space:')} select`,
						`${color.dim('Enter:')} confirm`,
						`${color.dim('Type:')} to search`,
					];

					// No results message
					const noResults =
						this.filteredOptions.length === 0 && userInput
							? [`${color.cyan(S_BAR)}  ${color.yellow('No matches found')}`]
							: [];

					const errorMessage =
						this.state === 'error' ? [`${color.cyan(S_BAR)}  ${color.yellow(this.error)}`] : [];

					// Get limited options for display
					const displayOptions = limitOptions({
						cursor: this.cursor,
						options: this.filteredOptions,
						style: (option, active) =>
							formatOption(option, active, this.selectedValues, this.focusedValue),
						maxItems: opts.maxItems,
						output: opts.output,
					});

					// Build the prompt display
					return [
						title,
						`${color.cyan(S_BAR)}  ${color.dim('Search:')} ${searchText}${matches}`,
						...noResults,
						...errorMessage,
						...displayOptions.map((option) => `${color.cyan(S_BAR)}  ${option}`),
						`${color.cyan(S_BAR)}  ${color.dim(instructions.join(' • '))}`,
						`${color.cyan(S_BAR_END)}`,
					].join('\n');
				}
			}
		},
	});

	// Return the result or cancel symbol
	return prompt.prompt() as Promise<Value | symbol>;
};
