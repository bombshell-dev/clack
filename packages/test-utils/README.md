# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
