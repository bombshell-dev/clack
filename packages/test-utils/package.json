{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
