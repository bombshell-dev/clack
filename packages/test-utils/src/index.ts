export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
