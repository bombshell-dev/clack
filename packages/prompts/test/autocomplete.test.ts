import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { autocomplete, autocompleteMultiselect } from '../src/autocomplete.js';
import { isCancel } from '../src/index.js';
import { MockReadable, MockWritable } from './test-utils.js';

describe('autocomplete', () => {
	let input: MockReadable;
	let output: MockWritable;
	const testOptions = [
		{ value: 'apple', label: 'Apple' },
		{ value: 'banana', label: 'Banana' },
		{ value: 'cherry', label: 'Cherry' },
		{ value: 'grape', label: 'Grape' },
		{ value: 'orange', label: 'Orange' },
	];

	beforeEach(() => {
		input = new MockReadable();
		output = new MockWritable();
	});

	afterEach(() => {
		vi.restoreAllMocks();
	});

	test('renders initial UI with message and instructions', async () => {
		const result = autocomplete({
			message: 'Select a fruit',
			options: testOptions,
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });
		await result;
		expect(output.buffer).toMatchSnapshot();
	});

	test('limits displayed options when maxItems is set', async () => {
		const options = [];
		for (let i = 0; i < 10; i++) {
			options.push({ value: `option ${i}`, label: `Option ${i}` });
		}

		const result = autocomplete({
			message: 'Select an option',
			options,
			maxItems: 6,
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });
		await result;
		expect(output.buffer).toMatchSnapshot();
	});

	test('shows no matches message when search has no results', async () => {
		const result = autocomplete({
			message: 'Select a fruit',
			options: testOptions,
			input,
			output,
		});

		// Type something that won't match
		input.emit('keypress', 'z', { name: 'z' });
		input.emit('keypress', '', { name: 'return' });
		await result;
		expect(output.buffer).toMatchSnapshot();
	});

	test('shows hint when option has hint and is focused', async () => {
		const result = autocomplete({
			message: 'Select a fruit',
			options: [...testOptions, { value: 'kiwi', label: 'Kiwi', hint: 'New Zealand' }],
			input,
			output,
		});

		// Navigate to the option with hint
		input.emit('keypress', '', { name: 'down' });
		input.emit('keypress', '', { name: 'down' });
		input.emit('keypress', '', { name: 'down' });
		input.emit('keypress', '', { name: 'down' });
		input.emit('keypress', '', { name: 'down' });
		input.emit('keypress', '', { name: 'return' });
		await result;
		expect(output.buffer).toMatchSnapshot();
	});

	test('shows selected value in submit state', async () => {
		const result = autocomplete({
			message: 'Select a fruit',
			options: testOptions,
			input,
			output,
		});

		// Select an option and submit
		input.emit('keypress', '', { name: 'down' });
		input.emit('keypress', '', { name: 'return' });

		const value = await result;
		expect(value).toBe('banana');
		expect(output.buffer).toMatchSnapshot();
	});

	test('shows strikethrough in cancel state', async () => {
		const result = autocomplete({
			message: 'Select a fruit',
			options: testOptions,
			input,
			output,
		});

		// Cancel with Ctrl+C
		input.emit('keypress', '\x03', { name: 'c', ctrl: true });

		const value = await result;
		expect(typeof value === 'symbol').toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});

	test('renders placeholder if set', async () => {
		const result = autocomplete({
			message: 'Select a fruit',
			placeholder: 'Type to search...',
			options: testOptions,
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });
		const value = await result;
		expect(output.buffer).toMatchSnapshot();
		expect(value).toBe('apple');
	});

	test('supports initialValue', async () => {
		const result = autocomplete({
			message: 'Select a fruit',
			options: testOptions,
			initialValue: 'cherry',
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });
		const value = await result;

		expect(value).toBe('cherry');
		expect(output.buffer).toMatchSnapshot();
	});

	test('can be aborted by a signal', async () => {
		const controller = new AbortController();
		const result = autocomplete({
			message: 'foo',
			options: testOptions,
			input,
			output,
			signal: controller.signal,
		});

		controller.abort();
		const value = await result;
		expect(isCancel(value)).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});
});

describe('autocompleteMultiselect', () => {
	let input: MockReadable;
	let output: MockWritable;
	const testOptions = [
		{ value: 'apple', label: 'Apple' },
		{ value: 'banana', label: 'Banana' },
		{ value: 'cherry', label: 'Cherry' },
		{ value: 'grape', label: 'Grape' },
		{ value: 'orange', label: 'Orange' },
	];

	beforeEach(() => {
		input = new MockReadable();
		output = new MockWritable();
	});

	afterEach(() => {
		vi.restoreAllMocks();
	});

	test('renders error when empty selection & required is true', async () => {
		const result = autocompleteMultiselect({
			message: 'Select a fruit',
			options: testOptions,
			required: true,
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });
		input.emit('keypress', '', { name: 'tab' });
		input.emit('keypress', '', { name: 'return' });
		await result;
		expect(output.buffer).toMatchSnapshot();
	});

	test('can be aborted by a signal', async () => {
		const controller = new AbortController();
		const result = autocompleteMultiselect({
			message: 'foo',
			options: testOptions,
			input,
			output,
			signal: controller.signal,
		});

		controller.abort();
		const value = await result;
		expect(isCancel(value)).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});
});
