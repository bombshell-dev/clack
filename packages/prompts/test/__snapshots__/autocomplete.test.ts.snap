// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`autocomplete > can be aborted by a signal 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m●[39m Apple
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > limits displayed options when maxItems is set 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select an option
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m●[39m Option 0
[36m│[39m  [2m○[22m [2mOption 1[22m
[36m│[39m  [2m○[22m [2mOption 2[22m
[36m│[39m  [2m○[22m [2mOption 3[22m
[36m│[39m  [2m○[22m [2mOption 4[22m
[36m│[39m  [2m...[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select an option
[90m│[39m  [2mOption 0[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > renders initial UI with message and instructions 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m●[39m Apple
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select a fruit
[90m│[39m  [2mApple[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > renders placeholder if set 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit
[36m│[39m
[36m│[39m  [2mSearch:[22m [2mType to search...[22m
[36m│[39m  [32m●[39m Apple
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select a fruit
[90m│[39m  [2mApple[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > shows hint when option has hint and is focused 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m●[39m Apple
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m○[22m [2mKiwi[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=3>",
  "<erase.down>",
  "[36m│[39m  [2mSearch:[22m
[36m│[39m  [2m○[22m [2mApple[22m
[36m│[39m  [32m●[39m Banana
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m○[22m [2mKiwi[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=5>",
  "<erase.down>",
  "[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [32m●[39m Cherry
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m○[22m [2mKiwi[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=6>",
  "<erase.down>",
  "[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [32m●[39m Grape
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m○[22m [2mKiwi[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=7>",
  "<erase.down>",
  "[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [32m●[39m Orange
[36m│[39m  [2m○[22m [2mKiwi[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=8>",
  "<erase.down>",
  "[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [32m●[39m Kiwi[2m (New Zealand)[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select a fruit
[90m│[39m  [2mKiwi[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > shows no matches message when search has no results 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m●[39m Apple
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=3>",
  "<erase.down>",
  "[36m│[39m  [2mSearch:[22m z█[2m (0 matches)[22m
[36m│[39m  [33mNo matches found[39m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=6>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select a fruit
[90m│[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > shows selected value in submit state 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m●[39m Apple
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=3>",
  "<erase.down>",
  "[36m│[39m  [2mSearch:[22m
[36m│[39m  [2m○[22m [2mApple[22m
[36m│[39m  [32m●[39m Banana
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select a fruit
[90m│[39m  [2mBanana[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > shows strikethrough in cancel state 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m●[39m Apple
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [2m○[22m [2mCherry[22m
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[31m■[39m  Select a fruit
[90m│[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocomplete > supports initialValue 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit
[36m│[39m
[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [2m○[22m [2mApple[22m
[36m│[39m  [2m○[22m [2mBanana[22m
[36m│[39m  [32m●[39m Cherry
[36m│[39m  [2m○[22m [2mGrape[22m
[36m│[39m  [2m○[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select a fruit
[90m│[39m  [2mCherry[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocompleteMultiselect > can be aborted by a signal 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo

[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [2m◻[22m Apple
[36m│[39m  [2m◻[22m [2mBanana[22m
[36m│[39m  [2m◻[22m [2mCherry[22m
[36m│[39m  [2m◻[22m [2mGrape[22m
[36m│[39m  [2m◻[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to navigate • [2mSpace:[22m[2m select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`autocompleteMultiselect > renders error when empty selection & required is true 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  Select a fruit

[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [2m◻[22m Apple
[36m│[39m  [2m◻[22m [2mBanana[22m
[36m│[39m  [2m◻[22m [2mCherry[22m
[36m│[39m  [2m◻[22m [2mGrape[22m
[36m│[39m  [2m◻[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to navigate • [2mSpace:[22m[2m select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[33m▲[39m  Select a fruit

[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [33mPlease select at least one item[39m
[36m│[39m  [2m◻[22m Apple
[36m│[39m  [2m◻[22m [2mBanana[22m
[36m│[39m  [2m◻[22m [2mCherry[22m
[36m│[39m  [2m◻[22m [2mGrape[22m
[36m│[39m  [2m◻[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to navigate • [2mSpace:[22m[2m select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=11>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[36m◆[39m  Select a fruit

[36m│[39m  [2mSearch:[22m [7m[8m_[28m[27m
[36m│[39m  [32m◼[39m Apple
[36m│[39m  [2m◻[22m [2mBanana[22m
[36m│[39m  [2m◻[22m [2mCherry[22m
[36m│[39m  [2m◻[22m [2mGrape[22m
[36m│[39m  [2m◻[22m [2mOrange[22m
[36m│[39m  [2m[2m↑/↓[22m[2m to navigate • [2mSpace:[22m[2m select • [2mEnter:[22m[2m confirm • [2mType:[22m[2m to search[22m
[36m└[39m",
  "<cursor.backward count=999><cursor.up count=10>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  Select a fruit
[90m│[39m  [2m1 items selected[22m",
  "
",
  "<cursor.show>",
]
`;
